app/uk/gov/hmrc/cdsreimbursementclaimfrontend/controllers/reimbursement/SelectDutyCodesController.scala

```diff
@@ -16,29 +16,186 @@
+                ) match {
+                  case (Some(_), Some(dutyCodesAnswer: DutyCodesAnswer)) =>
+                    dutyCodesAnswer.existsDutyTypeWithNoDutyCodesAnswer match {
+                      case Some(dutyType) =>
+                        Redirect(reimbursementRoutes.SelectDutyCodesController.showDutyCodes(dutyType))
+                      case None           => Redirect(reimbursementRoutes.EnterPaidAndClaimAmountController.start())
+                    }
+                  case _                                                 =>
+                    logger.warn("could not find duty types or duty codes")
+                    errorHandler.errorResult()
+                }
+            )
+        }
+    }
+  }
+
+  def showDutyCodes(dutyType: DutyType): Action[AnyContent] = authenticatedActionWithSessionData.async {
+    implicit request =>
+      withAnswers[DutyCodesAnswer] { (_, answer) =>
+        answer.fold(
+          Ok(selectDutyCodesPage(SelectDutyCodesController.selectDutyCodesForm(dutyType), dutyType))
+        )(dutyCodesAnswer =>
+          Ok(
+            selectDutyCodesPage(
+              SelectDutyCodesController.selectDutyCodesForm(dutyType).fill(dutyCodesAnswer),
+              dutyType
+            )
+          )
+        )
+      }
+  }
+
+  def submitDutyCodes(dutyType: DutyType): Action[AnyContent] =
+    authenticatedActionWithSessionData.async { implicit request: RequestWithSessionData[AnyContent] =>
+      withAnswers[DutyCodesAnswer] { (fillingOutClaim, maybeCurrentAnswer) =>
+        selectDutyCodesForm(dutyType)
+          .bindFromRequest()
+          .fold(
+            formWithErrors => BadRequest(selectDutyCodesPage(formWithErrors, dutyType)),
+            selectedAnswer => updateSessionCache(dutyType, selectedAnswer, maybeCurrentAnswer, fillingOutClaim)
+          )
+      }
+    }
+
+  private def updateJourneyStatus(
+    dutyTypesAnswer: DutyTypesAnswer,
+    dutyCodesAnswer: DutyCodesAnswer,
+    fillingOutClaim: FillingOutClaim
+  ): FillingOutClaim = {
+    val updatedReimbursementState = ReimbursementState.computeReimbursementState(
+      dutyTypesAnswer,
+      dutyCodesAnswer
+    )
+
+    FillingOutClaim.of(fillingOutClaim)(
+      _.copy(
+        dutyTypesSelectedAnswer = Some(updatedReimbursementState.dutyTypesAnswer),
+        dutyCodesSelectedAnswer = Some(updatedReimbursementState.dutyCodesAnswer)
+      )
+    )
+  }
+
+  private def updateSessionCache(
+    dutyType: DutyType,
+    dutyTypesSelectedAnswer: DutyCodesAnswer,
+    currentAnswer: Option[DutyCodesAnswer],
+    fillingOutClaim: FillingOutClaim
+  )(implicit hc: HeaderCarrier, request: RequestWithSessionData[AnyContent]): Future[Result] = {
+
+    val updatedDutyTypeToDutyCodesMap: Map[DutyType, List[TaxCode]] =
+      currentAnswer.fold(Map[reimbursement.DutyType, List[models.TaxCode]]()) { dutyCodesAnswer =>
+        dutyCodesAnswer.dutyCodes ++ List((dutyType -> dutyTypesSelectedAnswer.dutyCodes(dutyType)))
```

Comment:
This contains extra brackets
```
List(dutyType -> dutyTypesSelectedAnswer.dutyCodes(dutyType))
```

Reply:
Yeah, `scalafix` not working properly. I need to look into that. It should remove stuff like at compile time.

Reply:
done - manually :weary: